contracts/abstracts/TokenCollector.sol

```diff
@@ -31,87 +35,50 @@ abstract contract TokenCollector {
-        if (src == Source.Token) {
-            return _collectByToken(token, from, to, amount, srcData);
-        }
-        if (src == Source.TokenlonAllowanceTarget) {
-            return _collectByAllowanceTarget(token, from, to, amount);
-        }
-        if (src == Source.Permit2AllowanceTransfer) {
-            return _collectByPermit2AllowanceTransfer(token, from, to, amount, srcData);
-        }
-        if (src == Source.Permit2SignatureTransfer) {
-            return _collectByPermit2SignatureTransfer(token, from, to, amount, srcData);
-        }
-        revert("TokenCollector: unknown token source");
-    }
+        Source src = Source(uint8(data[0]));
 
-    function _collectByToken(
-        address token,
-        address from,
-        address to,
-        uint256 amount,
-        bytes memory data
-    ) private {
-        if (data.length > 0) {
-            (bool success, bytes memory result) = token.call(abi.encodePacked(IERC20Permit.permit.selector, data));
+        if (src == Source.TokenlonAllowanceTarget) {
+            return IAllowanceTarget(allowanceTarget).spendFromUserTo(from, token, to, amount);
+        } else if (src == Source.Token) {
+            return IERC20(token).safeTransferFrom(from, to, amount);
+        } else if (src == Source.TokenPermit) {
+            (bool success, bytes memory result) = token.call(abi.encodePacked(IERC20Permit.permit.selector, data[1:]));
             if (!success) {
                 assembly {
                     result := add(result, 0x04)
                 }
                 revert(abi.decode(result, (string)));
             }
-        }
-        IERC20(token).safeTransferFrom(from, to, amount);
-    }
-
-    function _collectByAllowanceTarget(
-        address token,
-        address from,
-        address to,
-        uint256 amount
-    ) private {
-        IAllowanceTarget(allowanceTarget).spendFromUserTo(from, token, to, amount);
-    }
-
-    function _collectByPermit2AllowanceTransfer(
-        address token,
-        address from,
-        address to,
-        uint256 amount,
-        bytes memory data
-    ) private {
-        require(amount < uint256(type(uint160).max), "TokenCollector: permit2 amount too large");
-        if (data.length > 0) {
-            (uint48 nonce, uint48 deadline, bytes memory permitSig) = abi.decode(data, (uint48, uint48, bytes));
-            IUniswapPermit2.PermitSingle memory permit = IUniswapPermit2.PermitSingle({
-                details: IUniswapPermit2.PermitDetails({ token: token, amount: uint160(amount), expiration: deadline, nonce: nonce }),
-                spender: address(this),
-                sigDeadline: uint256(deadline)
+            return IERC20(token).safeTransferFrom(from, to, amount);
+        } else if (src == Source.Permit2AllowanceTransfer) {
+            if (amount > uint256(type(uint160).max)) revert Permit2AmountTooLarge();
+            bytes memory permit2Data = data[1:];
```

Comment:
Can it also be `calldata`?

Reply:
It will be copied to memory when doing abi decode so I think no huge different here. Also it will cause stack too deep.